server-controller/hooks_test.go

```diff
@@ -29,104 +29,444 @@ import (
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClient{updateError: errorNew},
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Service deletion error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during service delete",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClient{deleteError: errorNew},
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Service creation error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during service create",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClient{createError: errorNew},
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:      "Ingress patch error",
+		oldServer: inferenceEndpoint,
+		newServer: inferenceEndpoint,
+		expected:  "ERROR during ingress update",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClient{updateError: errorNew},
+		},
+	},
+}
+
+func testUpdateTemplate(oldServer, newServer crv1.InferenceEndpoint, errs clientError) string {
+	var buf bytes.Buffer
+	log.SetOutput(&buf)
+	defer func() {
+		log.SetOutput(os.Stderr)
+	}()
+	updateMap := createTemplateClients(errs)
 	hooks := serverHooks{updateMap}
-	inferenceEndpoint := infer
-	hooks.Add(&inferenceEndpoint)
+	oldInferenceEndpoint := oldServer
+	newInferenceEndpoint := newServer
+	hooks.updateTemplate(&oldInferenceEndpoint, &newInferenceEndpoint)
 	return buf.String()
 }
 
+var patchDataTest = patchData{ModelName: "test", ModelVersionPolicy: "test", Namespace: "test", ResourcePath: "test"}
+
+var dataTestPrepareJSONPatchFromMap = []struct {
+	name          string
+	resourceType  string
+	mapPatch      []interface{}
+	oldData       map[string]string
+	newData       map[string]string
+	patchData     patchData
+	expectedError error
+	resourcePath  string
+}{
+	{
+		name:          "No changes",
+		resourceType:  "test",
+		mapPatch:      make([]interface{}, 0),
+		oldData:       map[string]string{"test": "test"},
+		newData:       map[string]string{"test": "test"},
+		patchData:     patchDataTest,
+		expectedError: nil,
+		resourcePath:  "",
+	},
+	{
+		name:          "Unequal maps",
+		resourceType:  "test",
+		mapPatch:      make([]interface{}, 0),
+		oldData:       map[string]string{"test": "test"},
+		newData:       map[string]string{"test1": "test"},
+		patchData:     patchDataTest,
+		expectedError: nil,
+		resourcePath:  "",
+	},
+	{
+		name:          "Wrong resource path",
```

Comment:
This case tests `fillTemplate` not `prepareJSONPatchFromMap`

Reply:
As we talked, I'm not sure it is worth effort to create separate tests for `fillTemplate`, as it would be hard to make falling `text/template`